pkg/util/winutil/eventlog/subscription/subscription.go

```diff
@@ -0,0 +1,455 @@
+// https://learn.microsoft.com/en-us/windows/win32/api/winevt/ne-winevt-evt_subscribe_flags
+func WithStartAfterBookmark(bookmark evtbookmark.Bookmark) PullSubscriptionOption {
+	return func(q *pullSubscription) {
+		q.SetBookmark(bookmark)
+	}
+}
+
+// WithStartAtOldestRecord will start the subscription from the oldest record in the event log.
+// https://learn.microsoft.com/en-us/windows/win32/api/winevt/ne-winevt-evt_subscribe_flags
+func WithStartAtOldestRecord() PullSubscriptionOption {
+	return func(q *pullSubscription) {
+		q.subscribeOriginFlag = evtapi.EvtSubscribeStartAtOldestRecord
+	}
+}
+
+// WithSubscribeFlags can be used to manually set EVT_SUBSCRIBE_FLAGS
+// https://learn.microsoft.com/en-us/windows/win32/api/winevt/ne-winevt-evt_subscribe_flags
+func WithSubscribeFlags(flags uint) PullSubscriptionOption {
+	return func(q *pullSubscription) {
+		q.subscribeFlags = flags
+	}
+}
+
+// WithNotifyEventsAvailable enables the EventsAvailable() channel
+func WithNotifyEventsAvailable() PullSubscriptionOption {
+	return func(q *pullSubscription) {
+		q.useNotifyChannel = true
+	}
+}
+
+func (q *pullSubscription) EventsAvailable() <-chan struct{} {
+	return q.notifyEventsAvailable
+}
+
+func (q *pullSubscription) Running() bool {
+	return q.started
+}
+
+func (q *pullSubscription) SetBookmark(bookmark evtbookmark.Bookmark) {
+	q.bookmark = bookmark
+	q.subscribeOriginFlag = evtapi.EvtSubscribeStartAfterBookmark
+}
+
+func (q *pullSubscription) Start() error {
+
+	if q.started {
+		return fmt.Errorf("Query subscription is already started")
+	}
+
+	var bookmarkHandle evtapi.EventBookmarkHandle
+	if q.bookmark != nil {
+		bookmarkHandle = q.bookmark.Handle()
+	}
+
+	// create subscription
+	hSubWait, err := newSubscriptionWaitEvent()
+	if err != nil {
+		return err
+	}
+	hSub, err := q.eventLogAPI.EvtSubscribe(
+		hSubWait,
+		q.channelPath,
+		q.query,
+		bookmarkHandle,
+		q.subscribeOriginFlag|q.subscribeFlags)
+	if err != nil {
+		safeCloseNullHandle(windows.Handle(hSubWait))
+		return err
+	}
+
+	// alloc reusable storage for EvtNext output
+	q.evtNextStorage = make([]evtapi.EventRecordHandle, q.eventBatchCount)
+
+	q.waitEventHandle = hSubWait
+	q.subscriptionHandle = hSub
+
+	if q.useNotifyChannel {
+		hStopWait, err := newStopWaitEvent()
+		if err != nil {
+			evtapi.EvtCloseResultSet(q.eventLogAPI, q.subscriptionHandle)
+			safeCloseNullHandle(windows.Handle(hSubWait))
+			return err
+		}
+		q.stopEventHandle = hStopWait
+
+		// Query loop management
+		q.notifyStop = make(chan struct{})
+		q.notifyNoMoreItems = make(chan struct{})
+		q.notifyNoMoreItemsComplete = make(chan struct{})
+		q.notifyEventsAvailable = make(chan struct{})
+		q.notifyEventsAvailableLoopDone = make(chan struct{})
+
+		// start goroutine to query events for channel
+		q.notifyEventsAvailableWaiter.Add(1)
+		go q.notifyEventsAvailableLoop()
+	}
+
+	q.started = true
+
+	return nil
+}
+
+func (q *pullSubscription) Stop() {
+	if !q.started {
```

Comment:
I'm not sure these methods are thread safe. In the tailer, this stop method can be called from the tailer's `Stop` method, or in the goroutine managing the event loop. It also appears it would be possible for `Start` to be called by the backoff/retry while stop is called at the same time by the tailer `Stop` method. 

I haven't analized the code clesely enough to see if this could actually happen at runtime - but I wanted to point it out regardless.

Reply:
Actually I think any race *should* be prevented by the done channel in the tailer. But feel free to take a closer look (or close this comment if it's irrelevant)

Reply:
You are right on both counts. `PullSubscription`'s exported methods don't support being called in parallel. The done channel forces the right order, `tailer.Stop` doesn't stop the subscription until `retry` stops and `tail` returns, then `tailer.Stop` is the only routine interacting with the subscription.